### DISCUSSION TITLE
SAML configuration with ADFS

### DISCUSSION THREAD

**Discussion by mastertIT** [kimai/kimai#4403] [General] [answered]
i looked at the Azure guide and it's near identical to what's needed for ADFS.

the claim types get outputted the same way.
the only exception is the groups have a transform so they don't output the group id directly. as result the group schema is http://schemas.xmlsoap.org/claims/Group

changed that in the group mapping and left the defaults as that was configured in adfs:

> mapping:            
> 
>             - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress, kimai: email }
>             - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name, kimai: username }
>             - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname, kimai: alias }
>             - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/displayname, kimai: title }
>         roles:
>             resetOnLogin: true
>             attribute: http://schemas.xmlsoap.org/claims/Group
>             mapping:
>                 # Insert your role-mapping here (ROLE_USER is added automatically)
>                 - { saml: Admin, kimai: ROLE_SUPER_ADMIN }
>                 - { saml: Manager, kimai: ROLE_ADMIN }
>                 - { saml: Teamlead, kimai: ROLE_TEAMLEAD }
> 

the redirection and adfs authentication flow works, however upon return to kimai we receive this error
![image](https://github.com/kimai/kimai/assets/94576286/716b72dc-3879-445d-a35f-9392b9bf1f9f)

ADFS side

![image](https://github.com/kimai/kimai/assets/94576286/dc9d93b1-ef97-4991-b80b-ed229aa15e61)

![image](https://github.com/kimai/kimai/assets/94576286/0fbca770-47f0-4e90-a179-b0c30d861fc6)

as far as i can tell a claim isn't being sent correctly? but unsure how to troubleshoot that as i don't see this in the tail logs

**Comment by kevinpapst** [comment]
You could try to add some logging: https://github.com/kimai/kimai/blob/main/src/Saml/SamlAuthenticator.php#L82

**Reply by mastertIT** [reply]
So to uncomment that line on that PHP file, reload cache, restart Apache? 
Try again and check the error?

**Reply by mastertIT** [reply]
[saml.xml.txt](https://github.com/kimai/kimai/files/13327688/saml.xml.txt)

attached xml saml post response from samlauthenticator.php

as far as we can tell the reply from adfs is correct and receiving the correct claim types.

is the role mapping claim expected to be in the same claim as the user?

**Reply by mastertIT** [reply]
after doing some more digging:

1. the role claim doesn't get passed in the initial post request but until the mapping is called.
if not called, according to the logic it should assign the default user role.

2. going through the onelogin saml guide there's one setting under the security class that needs to be set for ADFS, we tried enabling this as well but still receive the same error.

3. the baseurl is that just the url of the kimai domain? or should it resolve to a specific directory?, cannot find specific details on that.

4. can the Team be passed as a claim too? would that be: `kimai: Team_Teamname`
> 
kimai:
    saml:
        provider: microsoft
        activate: true
        title: Login with SSO
        mapping:            
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress, kimai: email }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name, kimai: username }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname, kimai: alias }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/displayname, kimai: title }
            #- { saml: $http://schemas.microsoft.com/ws/2008/06/identity/claims/role, kimai: team????
        roles:
            resetOnLogin: true
            attribute: http://schemas.xmlsoap.org/claims/Group
            mapping:
                # Insert your role-mapping here (ROLE_USER is added automatically)
                - { saml: Admin, kimai: ROLE_SUPER_ADMIN }
                - { saml: Manager, kimai: ROLE_ADMIN }
                - { saml: Teamlead, kimai: ROLE_TEAMLEAD }
        connection:
            idp:
                entityId: 'http://adfs.com/adfs/services/trust'
                singleSignOnService:
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                    url: 'https://adfs.com/adfs/ls/'
                singleLogoutService:
                    url: 'https://adfs.com/adfs/ls/'
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                x509cert: '=='
            # Your Kimai: replace https://www.example.com with your base URL
            sp:
                entityId: 'https://example.com/auth/saml/metadata'
                assertionConsumerService:
                    url: 'https://example.com/auth/saml/acs'
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'
                singleLogoutService:
                    url: 'https://example.com/auth/saml/logout'
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                #privateKey: ''
            # only set baseurl, if auto-detection doesn't work
            baseurl: 'https://example.com/auth/saml/acs'
            #strict: true
            debug: true
            security:
                nameIdEncrypted: false
                authnRequestsSigned: false
                logoutRequestSigned: false
                logoutResponseSigned: false
                wantMessagesSigned: false
                wantAssertionsSigned: false
                wantNameIdEncrypted: false
                requestedAuthnContext: false #true
                signMetadata: false
                #wantXMLValidation: true
                #signatureAlgorithm: 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'
                #digestAlgorithm: 'http://www.w3.org/2001/04/xmlenc#sha256'
                #ADFS URL-Encodes SAML data as lowercase, and the toolkit by default uses
                #uppercase. Turn it True for ADFS compatibility on signature verification
                #https://github.com/SAML-Toolkits/php-saml#how-it-works
                lowercaseUrlencoding: true #allow adfs compatibility

**Reply by mastertIT** [reply]
@kevinpapst any insight into this?
what else can we check if the claims are being returned correctly?

**Reply by kevinpapst** [reply]
I spent my full time building Kimai, but that does not mean that I have the capacity to solve every single post here (nor that I want to or). Pinging me to reach my inbox is annoying, actually not only for but for every maintainer once you reach a certain point of interest. I understand the urge, but try to understand what happens if dozens of people do it. Please only use it for emergency cases. 

If you want dedicated support: https://www.kimai.org/en/support.html

If you want me to debug your integration issue you find ways to contact me under the given link. 
And if you are already a customer then I need to know who you are.

**Reply by mastertIT** [reply]
i appreciate that very much. what i'm asking isn't for immediate support or an expectation of such.

what i asked was simply what other troubleshooting steps would you suggest based on the steps already taken.

**Comment by mastertIT** [comment, marked as answer]
Setup the Kimai Config first, as once saml is enabled you can use the metadata.xml to configure the ADFS side.
create in opt/kimai/config/packages/local.yaml

### Kimai Config:
Below Config Works with ADFS
```

kimai:
    saml:
        provider: microsoft
        activate: true 
        title: Login with ADFS
        mapping:            
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress, kimai: email }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress, kimai: username }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname, kimai: alias }
            - { saml: $http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name, kimai: title }
        roles:
            resetOnLogin: true
            attribute: $http://schemas.xmlsoap.org/claims/Group
            mapping:
                # Insert your role-mapping here (ROLE_USER is added automatically)
                - { saml: Admin, kimai: ROLE_SUPER_ADMIN }
                - { saml: Manager, kimai: ROLE_ADMIN }
                - { saml: Teamlead, kimai: ROLE_TEAMLEAD }
        connection:
            idp:
                entityId: 'https://adfs.com/adfs/services/trust'
                singleSignOnService:
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                    url: 'https://adfs.com/adfs/ls/'
                singleLogoutService:
                    url: 'https://adfs.com/adfs/ls/?wa=wsignout1.0'
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                x509cert: 'CERT'
            # Your Kimai: replace https://www.example.com with your base URL
            sp:
                entityId: 'https://kimaiurl.com/auth/saml/metadata'
                assertionConsumerService:
                    url: 'https://kimaiurl.com/auth/saml/acs'
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'
                singleLogoutService:
                    url: 'https://kimaiurl.com/auth/saml/logout'
                #    binding: 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'
                privateKey: ''
            # only set baseurl, if auto-detection doesn't work
            baseurl: 'https://kimaiurl.com/auth/saml/'
            strict: false
            debug: true
            security:
                nameIdEncrypted: true
                authnRequestsSigned: false
                logoutRequestSigned: false
                logoutResponseSigned: false
                wantMessagesSigned: false
                wantAssertionsSigned: false
                wantNameIdEncrypted: false
                requestedAuthnContext: true 
                signMetadata: false
                wantXMLValidation: true
                #signatureAlgorithm: 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256'
                #digestAlgorithm: 'http://www.w3.org/2001/04/xmlenc#sha256'
                #ADFS URL-Encodes SAML data as lowercase, and the toolkit by default uses
                #uppercase. Turn it True for ADFS compatibility on signature verification
                #https://github.com/SAML-Toolkits/php-saml#how-it-works
                lowercaseUrlencoding: true
            contactPerson:
                technical:
                    givenName: 'Name'
                    emailAddress: 'support@company.com'
            organization:
                en:
                    name: 'Company'
                    displayname: 'Name'
                    url: 'https://kimaiurl.com'
```


### ADFS Setup:
1. Add Relying Party Trust
![image](https://github.com/kimai/kimai/assets/94576286/0f0dc8a4-d290-42f9-ad9d-1a1193425507)

2. Select Data Source
![image](https://github.com/kimai/kimai/assets/94576286/237e664d-4b21-4d51-86e4-9196ffb51d43)

![image](https://github.com/kimai/kimai/assets/94576286/0b405937-cb22-42a0-a13b-04e482398d5f)

![image](https://github.com/kimai/kimai/assets/94576286/4a046a6e-f658-40b3-b2f7-e44e352a9686)

![image](https://github.com/kimai/kimai/assets/94576286/cd78faff-e0cf-445c-82fb-5af6ad241017)

3. Edit Claim Issuance Policy
![image](https://github.com/kimai/kimai/assets/94576286/fb743512-9f56-4032-af2d-dbe9448bc8ce)

4. Send LDAP Attributes as Claims:
![image](https://github.com/kimai/kimai/assets/94576286/e1793fcb-db23-4552-bec2-b825fce5e548)

![image](https://github.com/kimai/kimai/assets/94576286/5c731f93-e960-456c-a6cf-63df4f2cd5f4)

5. Send Group Memberships as a Claim:
![image](https://github.com/kimai/kimai/assets/94576286/6fd708af-798d-4f22-8f0f-b78668faeb0b)

![image](https://github.com/kimai/kimai/assets/94576286/71cf458d-aaf8-45b6-b3b5-403ed2de4487)

6. Apply Claim Issuance Policy:
![image](https://github.com/kimai/kimai/assets/94576286/c144134f-55d3-496e-b078-e77024b7c764)

Hit Finish and at this point both Kimai will use ADFS as it's SSO Provider.


Hoping to revise this once the SAML 'Teams' feature gets added #3846 

@kevinpapst : how can i add this guide to the SAML section of the documentation?

**Reply by kevinpapst** [reply]
Wow, thank you @mastertIT for sharing 👍 

So turns out ADFS means `Active Directory Federation Services` and is a Microsoft thing. No wonder I never heard about it 😁 

Would you create a new docs page OR add it as chapter to Azure AD?

In both cases we need a PR for the docs repo, the directory: https://github.com/kimai/www.kimai.org/tree/main/_documentation
And the image directory: https://github.com/kimai/www.kimai.org/tree/main/images/documentation

**Reply by mastertIT** [reply]
It should be it's own page, ADFS is the On-Premise Self Hosted version.
usually an Org. would use one or the other

**Reply by kevinpapst** [reply]
Ok, let me prepare that later on and then you could add some sentences around ?

**Reply by mastertIT** [reply]
Yeah Definitely.

**Reply by kevinpapst** [reply]
Here is the documentation: https://www.kimai.org/documentation/saml-adfs.html

And here you could change the docs: https://github.com/kimai/www.kimai.org/edit/main/_documentation/saml-adfs.md

**Reply by mastertIT** [reply]
i changed the order of the configuration and submitted a PR to match the workflow required to get it working. kimai needs to be configured first in order for the metadata.xml to be generated.
